File: 2021년/9월/0903 LCS.cpp
#include <iostream>
#include <cstring>
using namespace std;

string a,b;
int lcs[1001][1001];

int main()
{
	cin >> a >> b;
	for(int i=1;i<a.length()+1;i++)
	{
		for(int j=1;j<b.length()+1;j++)
		{
			if(a[i-1]==b[j-1])
			{
				lcs[i][j] = lcs[i-1][j-1] + 1;
			}
			else
			{
				lcs[i][j] = max(lcs[i][j-1], lcs[i-1][j]);
				
			}
		}
	}
	
	cout << lcs[-1][-1];
}

File: 2021년/9월/0903 LCS.py
# 백준_9251_LCS_DP_골드 5

a = input()
b = input()

a_len = len(a)
b_len = len(b)

lcs = [[0 for _ in range(b_len+1)]for _ in range(a_len+1)]

result = 0
for i in range(1, a_len+1):
    for j in range(1, b_len+1):
        if a[i-1] == b[j-1]:
            lcs[i][j] = lcs[i-1][j-1] + 1
            result = max(result, lcs[i][j])
        else:
            lcs[i][j] = max(lcs[i][j-1], lcs[i-1][j])

print(result)
